src/lib/rng/esdm_rng/esdm_rng.h

```diff
@@ -0,0 +1,107 @@
+#define ESDM_RNG_H
+
+#include <botan/rng.h>
+#include <mutex>
+
+namespace Botan {
+
+/**
+ * Return a shared reference to a global PRNG instance provided by ESDM
+ */
+BOTAN_PUBLIC_API(3, 6) RandomNumberGenerator& esdm_rng();
+
+/**
+ * Entropy Source and DRNG Manager (ESDM) is a Linux/Unix based
+ * PRNG manager, which can be seeded from NTG.1/SP800-90B sources.
+ *
+ * See:
+ *  - Repository: https://github.com/smuellerDD/esdm
+ *  - Further Docs: https://www.chronox.de/esdm/index.html
+ *
+ * Different entropy sources can be configured in respect of beeing
+ * active and how much entropy is accounted for each of them.
+ *
+ * ESDM tracks its seed and reseed status and blocks, when not fully seeded.
+ * For this functionality, the esdm_rpcc_get_random_bytes_pr (prediction resistant) or
+ * esdm_rpcc_get_random_bytes_full (fully seeded) calls have to be used.
+ *
+ * Configurable modes:
+ *   - fully seeded (-> fast): provide entropy from a DRBG/PRNG after beeing fully seeded,
+ *          block until this point is reached, reseed from after a time
+ *          and/or invocation limit, block again if reseeding is not possible
+ *   - prediction resistance (-> slow): reseed ESDM with fresh entropy after each invocation
+ *
+ * You typically want to use the fast fully seeded mode, which is the default.
+ *
+ * Instances of this class communicate over RPC with ESDM. The esdm_rpc_client
+ * library, provided by ESDM, is leveraged for this.
+ */
+class BOTAN_PUBLIC_API(3, 6) ESDM_RNG final : public Botan::RandomNumberGenerator {
+   public:
+      /**
+     * Default constructor for ESDM, fully seeded mode
+     */
+      ESDM_RNG();
+
+      /**
+      * Construct ESDM instance with configurable mode
+      *
+      * @param prediction_resistance: use prediction resistant mode if true,
+      *   fully seeded mode otherwise
+      */
+      explicit ESDM_RNG(bool prediction_resistance);
+
+      ~ESDM_RNG() override;
+
+      std::string name() const override;
+
+      /**
+      * ESDM blocks, if it is not seeded,
+      *
+      * @return true
+      */
+      bool is_seeded() const override;
+
+      /**
+      * ESDM can inject additional inputs
+      * but we do not account entropy for it
+      *
+      * @return true
+      */
+      bool accepts_input() const override;
+
+      void clear() override;
+
+   protected:
+      void fill_bytes_with_input(std::span<uint8_t> out, std::span<const uint8_t> in) override;
+
+   private:
+      /**
+      * tracks if predicition resistant or fully seeded interface should be queried
+      */
+      const bool m_prediction_resistance;
+
+      /**
+      * ESDM rpc client locks concurrent accesses, but initialization should be
+      * only done once
+      */
+      inline static std::mutex m_init_lock;
+
+      /**
+      * counts how many ESDM RNG instances are active in order to perform init and
+      * fini on first/last one
+      */
+      inline static size_t m_ref_cnt = 0;
```

Comment:
If we need a singleton, we should probably implement a singleton. 😏 For instance: implement an internal `ESDM_Context` class that lives in `esdm_rng.cpp`, and provides a static factory method instead of a public constructor. Instances of `ESDM_RNG` could then hold a shared pointer of this context. The factory method could hold a weak pointer to that instance, so that the `ESDM_Context` is destructed as soon as the last `ESDM_RNG` dies. Like so:

https://godbolt.org/z/nr8o5zEeq

Reply:
Note to self: look into the concurrency pitfalls of this....

Reply:
Re: concurrency: [cppreference.com](https://en.cppreference.com/w/cpp/memory/shared_ptr) states:

> All member functions (including copy constructor and copy assignment) can be called by multiple threads on different shared_ptr objects without additional synchronization even if these objects are copies and share ownership of the same object. If multiple threads of execution access the same shared_ptr object without synchronization and any of those accesses uses a non-const member function of shared_ptr then a data race will occur [...]

Additionally, [in std::weak_ptr<>::lock](https://en.cppreference.com/w/cpp/memory/weak_ptr/lock) it states:

> Effectively returns `expired() ? shared_ptr<T>() : shared_ptr<T>(*this)`, executed atomically.

Therefore, I am quite sure that [my proposed solution](https://godbolt.org/z/nr8o5zEeq) is correct, under the assumption that the "owned object" itself handles its thread synchronization as necessary (which your ESDM adapter is doing). The manipulation of the global weak ptr and in particular the lazy construction of the `ESDM_Context` object still need to be mutexed though. All shared_ptrs that come out of this factory method are copies of the weak_ptr's state though, so the condition stated in the first quote applies.

Reply:
I adapted your concept, but had to move weak_ptr and mutex to be class variables in order to use them in the destructor again.

Reply:
Thanks! Though, I think, the whole `weak_ptr` idea was a blunder, unfortunately. See here: https://github.com/randombit/botan/pull/4309#discussion_r1719738277